File: Cores/DotNet/Optimization/Algorithm.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using OSOL.Extremum.Cores.DotNet.Optimization.Nodes;

namespace OSOL.Extremum.Cores.DotNet.Optimization
{

    using Area = Dictionary<string, Tuple<double, double>>;

    public class Algorithm<TBase, TFuncType, TV> where TV : class, IOptimizable<TBase, TFuncType>
    {
        public State<TBase, TFuncType, TV> State = new State<TBase, TFuncType, TV>();
        public GeneralNode<TBase, TFuncType, TV> CurrentNode = null;

        public GeneralNode<TBase, TFuncType, TV>[] Nodes;
        public Tuple<int, int?, int>[] TransitionMatrix;

        public List<Func<object, JObject>> writers = new List<Func<object, JObject>>();

        public Algorithm(GeneralNode<TBase, TFuncType, TV>[] nodes, Tuple<int, int?, int>[] transitionMatrix)
        {
            this.Nodes = nodes;
            this.TransitionMatrix = transitionMatrix;
        }

        public void Initialize(Func<Dictionary<string, TFuncType>, TFuncType> f, Area area)
        {
            foreach (var n in Nodes)
            {
                n.Initialize(f, area, State);
            }

            CurrentNode = Nodes.First();
        }

        public void Reset()
        {
            this.State = new State<TBase, TFuncType, TV>();
            this.CurrentNode = null;
        }

        public TV Work(Func<Dictionary<string, TFuncType>, TFuncType> f, Area area, string logStates = null)
        {
            var logger = new Logger<TBase, TFuncType, TV>(logStates, writers);
            logger.Initialize();
            
            Initialize(f, area);
            logger.Log(State, CurrentNode);
            bool continueOrNot = true;
            while (continueOrNot)
            {
                CurrentNode.Process(f, area, State);
                logger.Log(State, CurrentNode);
                var currentConditionValue = CurrentNode.GetCurrentCondition(f, area, State);
                var nextNodes = TransitionMatrix
                    .Where(rule => rule.Item1 == CurrentNode.NodeId && rule.Item2 == currentConditionValue)
                    .Select(rule => (double) rule.Item3).ToArray();
                var nextNode = nextNodes.Length == 0 ? double.NaN : nextNodes[0];
                if (double.IsNaN(nextNode))
                {
                    continueOrNot = false;
                }
                else
                {
                    CurrentNode = Nodes.First(n => n.NodeId == (int) nextNode);
                }
            }
            logger.Log(State, CurrentNode);
            return State.result;
        }

        public JObject SerializeState() => this.State.ConvertToJson(writers);
    }

    public class Logger<TBase, TFuncType, TV> where TV : class, IOptimizable<TBase, TFuncType>
    {
        public int Counter = 1;

        public string LogLocation;
        public List<Func<object, JObject>> Writers;

        public Logger(string logLocation, List<Func<Object, JObject>> writers)
        {
            this.LogLocation = logLocation;
            this.Writers = writers;
        }

        public void PurgeFolder(DirectoryInfo dir)
        {
            foreach (FileInfo f in dir.GetFiles()) f.Delete();
            foreach (DirectoryInfo sd in dir.GetDirectories()) sd.Delete(true);
        }

        public void Initialize()
        {
            if (LogLocation != null)
            {
                var dir = new DirectoryInfo(LogLocation);
                if (dir.Exists)
                    PurgeFolder(dir);
                else
                    dir.Create();
            }
        }

        public void Log(State<TBase, TFuncType, TV> state, GeneralNode<TBase, TFuncType, TV> node)
        {
            if (LogLocation != null)
            {
                state.SetParameter("_nodeId", node.NodeId);
                var printer = new StreamWriter($"{LogLocation}/{Counter}.json");
                printer.WriteLine(state.ConvertToJson(Writers).ToString(Formatting.Indented));
                printer.Close();
                printer.Dispose();
                ++Counter;
            }
        }

    }
}

File: Cores/JVM/src/main/scala/OSOL/Extremum/Cores/JVM/Optimization/Algorithm.scala
package OSOL.Extremum.Cores.JVM.Optimization

import OSOL.Extremum.Cores.JVM.Optimization.Nodes.GeneralNode
import spray.json._
import java.io.{File, FileWriter}

import OSOL.Extremum.Cores.JVM.Optimization

final class Algorithm[Base, FuncType, V <: Optimizable[Base, FuncType]]
(nodes: Seq[GeneralNode[Base, FuncType, V]], transitionMatrix: Seq[(java.lang.Integer, Option[java.lang.Integer], java.lang.Integer)]) {

  var state: State[Base, FuncType, V] = new State()

  var currentNode: GeneralNode[Base, FuncType, V] = null

  var writers: Seq[Any => JsValue] = Seq.empty

  def reset(): Unit = {
    state = new State()
    currentNode = null
  }

  def initialize(f: Map[String, FuncType] => FuncType, area: Area): Unit = {
    nodes.foreach(_.initialize(f, area, state))
    currentNode = nodes.head
  }

  def work(f: Map[String, FuncType] => FuncType, area: Area, logStates: Option[String] = None): V = {
    val logger = new Optimization.Algorithm.Logger[Base, FuncType, V](logStates, writers)
    logger.initialize()

    initialize(f, area)
    logger(state, currentNode)
    var continue = true
    while (continue) {
      currentNode.process(f, area, state)
      logger(state, currentNode)
      val currentConditionValue = currentNode.getCurrentCondition(f, area, state)
      val nextNode = transitionMatrix.find { case (currentId, condition, _) =>
        currentId == currentNode.nodeId && condition == currentConditionValue }
      if (nextNode.isDefined) { currentNode = nodes.find(_.nodeId == nextNode.get._3).get }
      else { continue = false }
    }
    logger(state, currentNode)
    state.result.get
  }

  def serializeState(): JsValue = this.state.toJson(writers)

}

object Algorithm {

  case class Logger[Base, FuncType, V <: Optimizable[Base, FuncType]](logLocation: Option[String], writers: Seq[Any => JsValue]) {

    var counter = 1

    def purgeFolder(file: File): Unit = {
      if (file.isDirectory) file.listFiles.foreach(purgeFolder)
      if (file.exists && !file.delete) throw new Exception(s"Unable to delete ${file.getAbsolutePath}")
    }

    def initialize(): Unit = {
      if (logLocation.isDefined) {
        val location = new File(logLocation.get)
        if (location.exists()) purgeFolder(location)
        location.mkdir()
      }
    }

    def apply(state: State[Base, FuncType, V], node: GeneralNode[Base, FuncType, V]): Unit ={
      if (logLocation.isDefined){
        state.setParameter("_nodeId", node.nodeId)
        val printer = new FileWriter(s"${logLocation.get}/$counter.json")
        printer.write(state.toJson(writers).prettyPrint)
        printer.close()
        counter += 1
      }
    }
  }

}
